src/wallet/rpcwallet.cpp

```diff
@@ -3169,6 +3169,73 @@ UniValue generate(const JSONRPCRequest& request)
     return generateBlocks(coinbase_script, num_generate, max_tries, true);
 }
 
+UniValue rescanblockchain(const JSONRPCRequest& request)
+{
+    CWallet * const pwallet = GetWalletForJSONRPCRequest(request);
+    if (!EnsureWalletIsAvailable(pwallet, request.fHelp)) {
+        return NullUniValue;
+    }
+
+    if (request.fHelp || request.params.size() > 2) {
+        throw std::runtime_error(
+            "rescanblockchain (\"start_height\") (\"stop_height\")\n"
+            "\nRescan the local blockchain for wallet related transactions.\n"
+            "\nArguments:\n"
+            "1. \"start_height\"    (number, optional) block height where the rescan should start\n"
+            "2. \"stop_height\"     (number, optional) block height where the rescan should stop\n"
+            "\nResult:\n"
+            "{\n"
+            "  \"start_height\" :   (number) The block height where the rescan has started\n"
+            "  \"stop_height\" :    (number) The height of the last rescanned block\n"
+            "}\n"
+            "\nExamples:\n"
+            + HelpExampleCli("rescanblockchain", "\"100000 120000\"")
+            + HelpExampleRpc("rescanblockchain", "\"100000 120000\"")
+            );
+    }
+
+    LOCK2(cs_main, pwallet->cs_wallet);
+
+    CBlockIndex *pindexStart = chainActive.Genesis();
+    CBlockIndex *pindexStop = nullptr;
+    if (!request.params[0].isNull()) {
+        pindexStart = chainActive[request.params[0].get_int()];
+        if (!pindexStart) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid start_height");
+        }
+    }
+
+    if (!request.params[1].isNull()) {
+        pindexStop = chainActive[request.params[1].get_int()];
+        if (!pindexStop || pindexStop->nHeight < pindexStart->nHeight) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid stop_height");
+        }
+    }
+
+    // We can't rescan beyond non-pruned blocks, stop and throw an error
+    if (fPruneMode) {
+        CBlockIndex *block = pindexStop ?: chainActive.Tip();
+        while (block && block->nHeight > pindexStart->nHeight) {
+            if (!(block->nStatus & BLOCK_HAVE_DATA)) {
+                throw JSONRPCError(RPC_MISC_ERROR, "Can't rescan beyond pruned data. Use RPC call getblockchaininfo to determine your pruned height.");
+            }
+            block = block->pprev;
+        }
+    }
+
+    CBlockIndex *stopBlock = nullptr;
+    stopBlock = pwallet->ScanForWalletTransactions(pindexStart, pindexStop, true);
+    if (!stopBlock) {
+        // if we got a nullptr returned, ScanForWalletTransactions did rescan up to the requested stopindex
```

Comment:
minor bug: sadly, the comment about the return value for `ScanForWalletTransactions()` added in #10208 is incomplete. If `abortrescan` is called during a rescan, then `ScanForWalletTransactions()` will return `nullptr` even if the though scan didn't complete successfully to the tip (or requested `pindexStop` after this PR)

I think the only way to correctly return the status of the rescan would be to have more information passed back by `ScanForWalletTransactions()`.

(Test this by running `rescanblockchain` in one window and calling `abortrescan` in a different window. `rescanblockchain` returns the requested stop_height even though the rescan didn't extend that far).

@ryanofsky since he changed the semantics of `ScanForWalletTransactions()` in #10208.

Reply:
Nice catch... added a protection that detects aborted rescans via a check of  `CWallet::IsAbortingRescan()` after `ScanForWalletTransactions()`.

Reply:
Seems like a good pragmatic change. There's a (pathological) race condition where another thread starts a rescan and resets `fAbortRescan` before this RPC returns, but that's extremely unlikely.

I still think it would be nice to have `ScanForWalletTransactions()` give a more meaningful return value that indicates whether the rescan was aborted. The same bug exists in `importmulti` where `RescanFromTime()` will return the wrong time value if the rescan is aborted. However, that fix can be for another PR.